integrations/googledrive/src/actions.ts

```diff
@@ -1,81 +1,157 @@
+import { createActionWrapper } from '@botpress/common'
+import { RuntimeError } from '@botpress/sdk'
+import axios, { AxiosError } from 'axios'
+import { Stream } from 'stream'
 import { Client as DriveClient } from './client'
 import { wrapWithTryCatch } from './error-handling'
+import { FileChannelsCache } from './file-channels-cache'
+import { FileEventHandler } from './file-event-handler'
+import { FilesCache } from './files-cache'
+import { convertNormalToGeneric } from './validation'
 import * as bp from '.botpress'
 
-const listFiles: bp.IntegrationProps['actions']['listFiles'] = wrapWithTryCatch(async ({ client, ctx, input }) => {
-  const { nextToken } = input
-  const driveClient = await DriveClient.create({
-    client,
-    ctx,
-  })
-  return await driveClient.listFiles(nextToken)
-}, 'Error listing files')
+const injectToolsAndMetadata = createActionWrapper<bp.IntegrationProps>()({
+  toolFactories: {
+    driveClient: ({ client, ctx, logger }) => DriveClient.create({ client, ctx, logger }),
+    filesCache: async ({ client, ctx, logger }) => await FilesCache.load({ client, ctx, logger }),
+    fileChannelsCache: async ({ client, ctx, logger }) => await FileChannelsCache.load({ client, ctx, logger }),
+  },
+  extraMetadata: {
+    errorMessage: 'Placeholder error message', // TODO: Specify extraMetadata shape instead of providing placeholder values?
+  },
+})
 
-const listFolders: bp.IntegrationProps['actions']['listFolders'] = wrapWithTryCatch(async ({ client, ctx, input }) => {
-  const { nextToken } = input
-  const driveClient = await DriveClient.create({
-    client,
-    ctx,
-  })
-  return driveClient.listFolders(nextToken)
-}, 'Error listing folders')
+const wrapAction: typeof injectToolsAndMetadata = (meta, actionImpl) =>
+  injectToolsAndMetadata(meta, (props) =>
```

Comment:
**chore:** be careful with function currying and injection, this is hard to read and I'd prefer the dummy way of doing things